sharktank/sharktank/utils/testing.py

```diff
@@ -108,6 +110,234 @@ def assertFileWritten(self, p: Path):
         self.assertGreater(p.stat().st_size, 0, msg=f"Expected file {p} had zero size")
 
 
+class IreeVsEagerLLMTest(TempDirTestBase):
+    """Class for comparing the results of IREE and eager execution of the same LLM."""
+
+    def compare_outputs(
+        self,
+        *,
+        eager_result: torch.Tensor,
+        iree_result: torch.Tensor,
+        stage_name: str,
+        rtol: float | None = None,
+        atol: float | None = None,
+    ) -> None:
+        """
+        Compare the iree results with the eager results, one sequence at a time.
+
+        Args:
+            eager_result: The result from the eager execution
+            iree_result: The result from the IREE execution
+            stage_name: The name of the stage being compared (e.g., "prefill", "decode")
+            rtol: Relative tolerance for the comparison
+            atol: Absolute tolerance for the comparison
+        """
+        for i, (eager_i, iree_i) in enumerate(zip(eager_result, iree_result)):
+            try:
+                torch.testing.assert_close(
+                    actual=iree_i, expected=eager_i, rtol=rtol, atol=atol
+                )
+            except AssertionError as error:
+                raise AssertionError(
+                    f"Outputs do not match for {stage_name} batch index {i}:\n"
+                    f"Eager: {eager_i}\n"
+                    f"IREE: {iree_i}\n"
+                ) from error
+
+    def inplace_transfer_theta(self, node: Theta | dict | AnyTensor, dev: torch.device):
+        """
+        Inplace transfer all of the weights in the given Theta to the specified torch.device to allow for gpu-based eager tests regardless of how the initial weights were created.
+        """
+        if isinstance(node, AnyTensor):
+            return node.to(dev)
+
+        if isinstance(node, dict):
+            for key in node.keys():
+                node[key] = self.inplace_transfer_theta(node[key], dev)
+            return node
+
+        for key in node.keys:
+            self.inplace_transfer_theta(node.tensor(key), dev)
```

Comment:
It seems that the current theta would not change as when a new tensor gets returned our theta would not get updated.

Reply:
node.tensor(key) gets updated in place